staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"time"
+
+	"github.com/google/cel-go/common/types/ref"
+
+	v1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/generic"
+)
+
+// Validation contains the inputs needed to compile, evaluate and validate a cel expression
+type Validation struct {
+	Expression string
+	Message    string
```

Comment:
`Message` and `Reason` only make sense for validation, not matching.

I'm wondering if what we're looking at for "evaluator" is actually more of a filter function, where the CEL expression is the filter key.

Something like:

```
type CELFilterable interface {
   Expression() string
}

// For Validation:

// add Expression() method to the Evaluation{} struct

ValidationFilter, err := filterCompiler.Compile([]Evaluation}{}) // test err

Validations, err := ValidationFilter.ForInput(request) // err results in failure policy here

for _, result := range Validations {
   // collate results
}

// For matching:

matchExpressions := someVal // taken from the resource

matcher, err := filterCompiler.Compile(matchExpressions) // test err

matches, err := matcher.ForInput(request) // err results in failure policy here

if len(matches) != len(matchExpressions) {
  // one of the match expressions did not match, end evaluation
}
```

Doing things this way decouples the notion of applying CEL embedded in an object from the syntactic significance of the result (e.g. nothing in the above requires match logic to maintain a stub implementation of violation messages).

One downside of this approach is you lose context for expressions that do not match. This is currently moot for both validation and matching. If we decide we want that, then matcher.ForInput could act as a sifter:

`func ForInput(request) (matching, nonMatching []CELFilterable, error)`

Reply:
The caller will have to do some type conversion, unless we want to play around with generics

Reply:
I agree with this direction, although I was confused by the name `CELFilterable` (specifically the filterable part). How about something like `ExpressionAccessor` instead?

Reply:
That works! What do you think about whether to use generics to avoid requiring the caller do type conversion?

```golang

validationFilter, err := filter.Compile[Validation]([]Validation{})

res, err := validationFilter.ForInput(request)
```

(also not super opinionated on `filter` vs. `filterCompiler` vs. some better name)

Reply:
It looks like validation's logic is: "if CEL statement is not satisfied, throw violation"

IIRC match's logic is: "if match CEL expression not satisfied, no match"

Which are opposite sides of the same coin. There are a few options for this:

1) The sifter model is one way to bridge the gap.

2) We could also have it be an inverse filter (return non-matched objects), and match just makes sure the length of that list is zero. 

3) We could add a toggle to the function call to say whether we're filtering or not.

I think (2) is simplest.

Reply:
Decoupled, I have Matcher and validator defined as 

```
// Validator is contains logic for converting ValidationEvaluation to PolicyDecisions
type Validator interface {
	Filter
	// Validate is used to take cel evaluations and convert into decisions
	Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) []PolicyDecision
}

// Matcher is contains logic for converting MatchEvaluation to bool of matches or does not match
type Matcher interface {
	Filter
	// Match is used to take cel evaluations and convert into decisions
	Match(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) bool
}
```

And they'll fully encapsulate the decision making logic